### PR TITLE
Png loader example

### PR DESCRIPTION
**Pull request by slimbuck**
This PR adds to the engine an example texture parser for loading PNGs manually and adds an accompanying example.

Manual loading of PNG is useful under iOS which still tramples on texture alpha (see #1930).

The example is completely static, just demonstrating a lightmap without banding and probably needs more work before merging:

Screenshot 2021-08-20 at 16 57 56

### CODE REVIEW

File: examples/src/examples/loaders/png.tsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import * as pc from 'playcanvas/build/playcanvas.js';
+import Example from '../../app/example';
+
+class OBJExample extends Example {
```
Comment (mvaligursky) [inline examples/src/examples/loaders/png.tsx:4]: rename the class

**Comment by yaustar** [comment]
For editor only projects, this is only possible when self hosting as the developer would have to hook in the parser between the application creation and config load function call in `__start__.js`

**Comment by yaustar** [comment]
Example of adding the upng loader to a self hosted solution:

Add `upng.js` to the project folder:
image

Include it in the HTML
image

Add the parser to the texture loader in `__start__.js`
```
try {
        app = new pc.Application(canvas, {
            elementInput: devices.elementInput,
            keyboard: devices.keyboard,
            mouse: devices.mouse,
            gamepads: devices.gamepads,
            touch: devices.touch,
            graphicsDeviceOptions: window.CONTEXT_OPTIONS,
            assetPrefix: window.ASSET_PREFIX || "",
            scriptPrefix: window.SCRIPT_PREFIX || "",
            scriptsOrder: window.SCRIPTS || []
        });

        app.loader.getHandler("texture").parsers.png = new PngParser(app.graphicsDevice);
    }
```

File: scripts/parsers/upng.js [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+    this._device = device;
+}
+
+Object.assign(PngParser.prototype, {
+    load: function (url, callback, asset) {
+        pc.Asset.fetchArrayBuffer(url.load, callback, asset, this.maxRetries);
+    },
+
+    open: function (url, data, device) {
+        const img = UPNG.decode(data);
```
Comment (willeastcott) [inline scripts/parsers/upng.js:439]: Can you just do this JS decode if the rgbm flag is set? And use the browser PNG loader otherwise?

**Comment by yaustar** [comment]
Just thought of a way to patch it into the engine: https://playcanvas.com/project/870259/overview/added-upng

This adds the upng to the project and an after engine patch on the addHandler function that changes the PNG parser

**Pull request closed without merging**